### PR TITLE
Add PBKDF2 to ACVP Testing Harness

### PR DESCRIPTION
**Pull request by billbo-yang**
### Issues:
Resolves [CryptoAlg-1271](https://sim.amazon.com/issues/CryptoAlg-1271)

### Description of changes: 
Currently AWS-LC has PBKDF2 within the FIPS boundary, but doesn't support testing it with it's ACVP tool. This PR adds support for PBKDF2 to the ACVP tool.

### Call-outs:
No architecture or design changes.

### Testing:
Change tested by running the ACVP tool on PBKDF2 in online mode to fetch the appropriate vectors, compute the result, then upload said results to be compared to the expected results by NIST. Also tested by adding PBKDF support for the ACVP tool testing scripts and doing the same process as above locally.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: util/fipstools/acvp/acvptool/subprocess/pbkdf.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2020, Google Inc.
```
Comment (torben-hansen) [inline util/fipstools/acvp/acvptool/subprocess/pbkdf.go:1]: New file?

Use AWS license then :)  Example: https://github.com/awslabs/aws-lc/blob/main/crypto/mem_test.cc#L1-L2

File: util/fipstools/acvp/modulewrapper/modulewrapper.cc [hunk 2]
```diff
@@ -393,6 +393,46 @@ static bool GetConfig(const Span<const uint8_t> args[], ReplyCallback write_repl
...
+            ],
+            "keyLen": [
+              {
+                "min": 112,
+                "max": 2048,
+                "increment": 8
+              }
+            ],
+            "hmacAlg": [
+              "SHA2-224"
```
Comment (Taffer) [inline util/fipstools/acvp/modulewrapper/modulewrapper.cc:431]: Should this list match the list of supported hashes?

File: util/fipstools/acvp/acvptool/subprocess/pbkdf.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+// List of supported HMAC Algorithms by NIST
+var HmacAlgList map[string]bool = map[string]bool{
+	"SHA-1":        true,
+	"SHA2-224":     true,
+	"SHA2-256":     true,
+	"SHA2-384":     true,
+	"SHA2-512":     true,
+	"SHA2-512/256": true,
```
Comment (Taffer) [inline util/fipstools/acvp/acvptool/subprocess/pbkdf.go:63]: SHA2-512/256 isn't Approved in FIPS mode (see `PBKDF2_verify_service_indicator()` in `crypto/fipsmodule/service_indicator/service_indicator.c`).

**Review by Taffer — APPROVED** [review]
Looks good to me!

File: util/fipstools/acvp/modulewrapper/modulewrapper.cc [hunk 4]
```diff
@@ -2058,6 +2102,36 @@ static bool FFDH(const Span<const uint8_t> args[], ReplyCallback write_reply) {
...
+  const Span<const uint8_t> hmac_name = args[3];
+  const Span<const uint8_t> key_len = args[4];
+
+  // read bit data into useable variables
+  unsigned int iterations_uint;
+  memcpy(&iterations_uint, iterations.data(), sizeof(iterations_uint));
+  unsigned int key_len_uint;
+  memcpy(&key_len_uint, key_len.data(), sizeof(key_len_uint));
+
+  key_len_uint = key_len_uint/8;
```
Comment (torben-hansen) [inline util/fipstools/acvp/modulewrapper/modulewrapper.cc:2118]: Any restrictions in key length we should catch here?

File: util/fipstools/acvp/acvptool/subprocess/pbkdf.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+				return nil, fmt.Errorf("test case %d/%d contains password %s of length %d, but expected password length between %d and %d", group.ID, test.ID, test.Pwd, len(test.Pwd), passwordLenMin, passwordLenMax)
+			}
+
+			if test.Iterations > iterationMax || test.Iterations < iterationMin {
+				return nil, fmt.Errorf("test case %d/%d contains iteration count of %d, but expected iteration count between %d and %d", group.ID, test.ID, test.Iterations, iterationMin, iterationMax)
+			}
+
+			testResp := pbkdfTestResponse{ID: test.ID}
+
+			// encode hex
```
Comment (torben-hansen) [inline util/fipstools/acvp/acvptool/subprocess/pbkdf.go:118]: Decode I guess?

Comment is also sorta redundant. It is obvious what is going on.

File: util/fipstools/acvp/acvptool/subprocess/pbkdf.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+			var result [][]uint8
+			result, err := m.Transact("PBKDF", 1, []byte(test.Pwd),
+				salt, uint32le(uint32(test.Iterations)), []byte(group.HmacAlg),
+				uint32le(uint32(test.KeyLen)))
+			if err != nil {
+				return nil, fmt.Errorf("PBKDF operation failed: %s", err)
+			}
+
+			// debug putting all letterse to upper in the output for now
```
Comment (torben-hansen) [inline util/fipstools/acvp/acvptool/subprocess/pbkdf.go:132]: Not sure I get what you mean here?

File: util/fipstools/acvp/acvptool/subprocess/pbkdf.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+		}
+
+		// Check each test in the test group
+		for _, test := range group.Tests {
+
+			if test.KeyLen > keyLenMax || test.KeyLen < keyLenMin {
+				return nil, fmt.Errorf("test case %d/%d requests output key of length %d, but expected output key length between %d and %d", group.ID, test.ID, test.KeyLen, keyLenMin, keyLenMax)
+			}
+
+			// check length of salt against NIST min/max
```
Comment (torben-hansen) [inline util/fipstools/acvp/acvptool/subprocess/pbkdf.go:103]: Just to nitpick pedantically: Start sentences with capital letter :)

File: util/fipstools/acvp/modulewrapper/modulewrapper.cc (hunk 4 above)
Comment (billbo-yang) [reply util/fipstools/acvp/modulewrapper/modulewrapper.cc:2118]: Restrictions in key length are caught in [line 99](https://github.com/awslabs/aws-lc/pull/657/files#diff-5ad898a0522db58a0c3e9e6084109fde76013892781d282bc86b86e17afc3a52R99) of `pbkdf.go`, which is where key_len is obtained from - so I don't think its necessary here.

File: util/fipstools/acvp/acvptool/subprocess/pbkdf.go (hunk 6 above)
Comment (billbo-yang) [reply util/fipstools/acvp/acvptool/subprocess/pbkdf.go:132]: Was an old debug statement that slipped through the cracks, removed :)

**Review by torben-hansen — APPROVED** [review]
LGTM

File: util/fipstools/acvp/acvptool/subprocess/pbkdf.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
```
Comment (samuel40791765) [inline util/fipstools/acvp/acvptool/subprocess/pbkdf.go:2]: License change update PR was just merged, so new files will have to add `OR ISC` 👍 
```suggestion
// SPDX-License-Identifier: Apache-2.0 OR ISC
```

**Review by darylmartin100 — APPROVED** [review]

**Pull request merged**